Update from-scratch networking section, include NetworkPolicy

caseydavenport opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/kubernetes/kubernetes.github.io/1174)
<!-- Reviewable:end -->

caseydavenport commented:
Not sure what is going on with the CLA bot, I've definitely got CLA approval in the past :)

caseydavenport commented:
Thanks @googlebot!

devin-donnelly commented on docs/getting-started-guides/scratch.md:
```diff
@@ -66,21 +67,25 @@ the node is added.  A process in one pod should be able to communicate with
+      - Manually configure static routing external to Kubernetes.
   - Generally highest performance.
-- Create an Overlay network
+- Use a network plugin 
```
As this is one of the two major methods, I'd format it as bold text so that it's more obvious.

devin-donnelly commented on docs/getting-started-guides/scratch.md:
```diff
@@ -66,21 +67,25 @@ the node is added.  A process in one pod should be able to communicate with
 accomplished in two ways:
 
-- Configure network to route Pod IPs
+- Configure underlay network to route Pod IPs
```
As this is one of the two major methods, I'd format it as bold text so that it's more obvious.

devin-donnelly commented:
Just two minor issues. Also, is this intended for the 1.4 docs or the current live docs?

Either way, please rebase it against the latest target branch.

caseydavenport commented:
@devin-donnelly Thanks, I've updated as per your comments and rebased against master.

This doesn't have any v1.4 specific changes, so it should be against the current live docs.

devin-donnelly commented:
Docs LGTM. @erictune , can you do a Tech Review on this?

erictune commented on docs/getting-started-guides/scratch.md:
```diff
@@ -116,6 +121,16 @@ Also, you need to pick a static IP for master node.
+
+Kubernetes enables the definition of fine-grained network policy between Pods 
+using the [NetworkPolicy](/docs/user-guide/networkpolicy) resource.
+
```
I assume not all network plugins support NetworkPolicy?  State this explicitly.

erictune commented on docs/getting-started-guides/scratch.md:
```diff
@@ -66,21 +67,25 @@ the node is added.  A process in one pod should be able to communicate with
-  - Need to make the Pod IPs routable by programming routers, switches, etc.
-  - Can be configured external to Kubernetes, or can implement in the "Routes" interface of a Cloud Provider module.
+  - Need to make the Pod IPs routable by programming routers, switches, etc.  
+    - This can be done in a few different ways:
```
Can this line fold into previous and then lose a level of indent?

erictune commented on docs/getting-started-guides/scratch.md:
```diff
@@ -66,21 +67,25 @@ the node is added.  A process in one pod should be able to communicate with
-  - Can be configured external to Kubernetes, or can implement in the "Routes" interface of a Cloud Provider module.
+  - Need to make the Pod IPs routable by programming routers, switches, etc.  
+    - This can be done in a few different ways:
+      - Implement in the "Routes" interface of a Cloud Provider module.
```
(Requires building your own fork of Kubernetes or contributing code to the project, right?)

caseydavenport replied:
yes, I think so.

erictune commented on docs/getting-started-guides/scratch.md:
```diff
@@ -66,21 +67,25 @@ the node is added.  A process in one pod should be able to communicate with
+      - Manually configure static routing external to Kubernetes.
   - Generally highest performance.
-- Create an Overlay network
+- **Use a network plugin**
```
It seems like there are two axes here, getting crammed into one:

1. where support lives
   1. in a script or person who statically programs your network hardware, and has no direct interaction with kubernetes.  You have to write this yourself, but there are no rules.
   1. in a plugin which is a separate piece of software that Kubernetes calls out to, called a Network Plugin.  You can use an existing one or write your own.  You need to go find one somewhere other than on the Kubernetes website.
   1. support compiled directuly into Kubernetes (if you don't see one you like, you'd need to write some go code, and then building your own fork of Kubernetes or send us a PR). 
1. How the packets are made to go to the Pod IPs
  - by distributing routes (== underlay?)
  - by encapsulation (== overlay ?)

caseydavenport commented:
OK - The latest commit reworks this according to @erictune's suggestions.  I think I like this quite a bit more (thanks @erictune!)  - let me know if you think it's better and I will squash.

CC @devin-donnelly for docs review.

devin-donnelly commented:
Fine by me. I'll merge it in.

erictune commented:
thanks @caseydavenport looks good.
